test/TensorFlowNET.UnitTest/Training/GradientDescentOptimizerTests.cs

```diff
@@ -115,5 +166,72 @@ public void TestTensorLearningRate()
             TestTensorLearningRate<float>();
             TestTensorLearningRate<double>();
         }
+
+        public void TestGradWrtRef<T>() where T : struct
+        {
+            var dtype = GetTypeForNumericType<T>();
+
+            var graph = tf.Graph().as_default();
+            using (var sess = self.cached_session())
+            {
+                var opt = tf.train.GradientDescentOptimizer(3.0f);
+                var values = new[] { 1.0, 3.0 };
+                var vars_ = values.Select(
+                        v => tf.Variable(new[] { v }, dtype: dtype) as IVariableV1
+                    ).ToList();
+                var grads_and_vars = opt.compute_gradients(tf.add(vars_[0], vars_[1]), vars_);
+                sess.run(tf.global_variables_initializer());
+                foreach (var (grad, _) in grads_and_vars)
+                    self.assertAllCloseAccordingToType(new[] { 1.0 }, self.evaluate<T[]>(grad));
+
+            }
+        }
+
+        [TestMethod]
+        public void TestGradWrtRef()
+        {
+            TestGradWrtRef<float>();
+            TestGradWrtRef<double>();
+        }
+
+        public void TestWithGlobalStep<T>() where T : struct
+        {
+            var dtype = GetTypeForNumericType<T>();
+
+            tf.Graph().as_default();
+            using (var sess = self.cached_session())
+            {
+                var global_step = tf.Variable(0, trainable: false);
+                var var0 = tf.Variable(new[] { 1.0, 2.0 }, dtype: dtype);
+                var var1 = tf.Variable(new[] { 3.0, 4.0 }, dtype: dtype);
+                var grads0 = tf.constant(new[] { 0.1, 0.1 }, dtype: dtype);
+                var grads1 = tf.constant(new[] { 0.01, 0.01 }, dtype: dtype);
+                var grads_and_vars = new[] {
+                    Tuple.Create(grads0, var0 as IVariableV1),
+                    Tuple.Create(grads1, var1 as IVariableV1)
+                };
+                var sgd_op = tf.train.GradientDescentOptimizer(3.0f)
+                    .apply_gradients(grads_and_vars, global_step: global_step);
```

Comment:
@AsakusaRinne why does `apply_gradiens` take `System.Tuple` while `zip` produces `(T1, T2)`. Would it be better to replace or extend `apply_gradients` interface to support valuetuple as well?